lib/code-caller/code-caller-container.js

```diff
@@ -319,6 +316,25 @@ class CodeCallerContainer {
     this.debug('enter ensureChild()');
     this._checkState();
 
+    await this._ensureChild();
+
+    // Give the child a reasonably long amount of time to start up. Under
+    // periods of high load, it should be cheaper to wait for a child to
+    // start slowly than it would be to time out too quickly and try to
+    // launch a new one to replace it.
+    await this.call('ping', null, null, 'ping', [], { timeout: 60 * 1000 });
+
+    this._checkState();
+    this.debug('exit ensureChild()');
+  }
+
+  /**
+   * Ensures that this code caller has a container.
+   */
+  async _ensureChild() {
```

Comment:
Rather than having `_ensureChild()` and `ensureChild()` could we just have a single `ensureChild()` function and do the ping after `_createAndAttachContainer()`?

Reply:
I don't think so, unfortunately. Because `call()` needs to call some form of `ensureChild()` as it's booting up, we need to differentiate "just create the container" and "ensure the container and health-check it". Otherwise, `ensureChild()` would call `call('ping')`, which would then call `ensureChild()` again 😬 

I suppose we could have `call()` skip the call to `ensureChild()` if it's invoked with `ping`?

Reply:
Alternatively, now that the pool of code callers always invokes `ensureChild()` after it creates a worker and before it leases it out, we could probably drop the internal calls to `ensureChild`? For that matter, I think `prepareForCourse` always calls `ensureChild()`, so a call to `ensureChild()` inside `call()` shouldn't be necessary for at least those two reasons.

Reply:
Oh right, I was forgetting that `ping` calls `call()`! I would prefer a `doNotPing` argument to `ensureChild()` rather than having two versions (`_ensureChild()` and `ensureChild()`).

Remind me how the state machine works here -- will `ensureChild()` ever recreate a new child (maybe the child died?), or is this a one-and-done thing the first time we create a child? If the latter then I agree that dropping all the internal `ensureChild()` calls would be the cleanest (it would actually clean up the code in general) and we could just unconditionally create a child on initialization.

Reply:
I changed `code-caller-container` to no longer use an internal call to `ensureChild()` in b828127. I made a similar change to `code-caller-native` in 1334a5e; `_startChild()` is no longer called from `call()`, and the caller must be in state `WAITING` when a call is made. This is consistent with how `code-caller-container` works.

Reply:
As for how the state machine works, we can never transition from `EXITING` or `EXITED` back to `CREATED` or `WAITING`; as you put it, we'll never recreate the container or child process. If we need a new container or child process, the pool will detect that and create an entirely new `CodeCaller{Native,Container}` object.

Reply:
Great! This feels much cleaner.